### PR TITLE
Adds Pat Walter's Chembl Filters extraction to the FilterCatalog

### PR DESCRIPTION
**Pull request by bp-kelley**
I need to add some testing, but there are some smarts patterns we should probably resolve that throw warnings in merge Hs due to patterns like:

```
[!#7,#1]~[NX2R0]=[NX2R0]~[!#7,#1]
```

That seem a bit superfluous.

### CODE REVIEW

**Comment by greglandrum** [comment]
@bp-kelley please remove the WIP tag and ping me when you're ready to have this reviewed.

**Comment by bp-kelley** [comment]
There are some queries from taken from CHEMBL that don't fit the rdkit hydrogen model, PatW and I are fixing these and will remove the WIP when resolved.

**Comment by bp-kelley** [comment]
@greglandrum , @PatWalters Here are the mergeHs queries incompatible queries that I changed, please take a look when you have time.

Changes:
   acyclic imine:  
        C-!@[NX2]=[C!R,#1]-C -> 
        C-!@[NX2]=[C!R]-C
        hypervalent hydrogen dropped
   
   Filter5_azo:
        [!#7,#1]~[NX2R0]=[NX2R0]~[!#7,#1] =>
        [!#7]~[NX2R0]=[NX2R0]~[!#7]         
        X is explicit connections, don't need explicit hydrogen
   
   Filter18_oxime_ester:
       [!#7,#1]C=NO[C,S,P](=O)* ->    
       [!#7][CX3]=NO[C,S,P](=O)*
       Adding explicit connections takes care of #1
   
   Filter30_beta_halo_carbonyl:
        [#6,#1]C(=O)CC[Br,I] ->
        [$(C(-C)(-C)(=O)),$([CH](=O)-C)](=O)CC[Br,I]
        Make explicit recursive smarts for the #6,#1

**Comment by PatWalters** [comment]
acyclic imine:
***@***.***=[C!R,#1 <https://github.com/rdkit/rdkit/issues/1>]-C ->
***@***.***=[C!R]-C
hypervalent hydrogen dropped

I think the intent of the acyclic imine rule was to catch both and
secondary imines.   The revised rule doesn't catch primary imines.   This
seems to do the trick.  There may be a simpler way to do this.

***@***.***)

Filter5_azo:
[!#7 <https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>][NX2R0]=[NX2R0][!#7
<https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>] =>
[!#7 <https://github.com/rdkit/rdkit/issues/7>][NX2R0]=[NX2R0][!#7
<https://github.com/rdkit/rdkit/issues/7>]
X is explicit connections, don't need explicit hydrogen

In this case the intent appears to be to also catch things like
CN=N
Which would be missed by the revised pattern.  I think I'd write that one
as

[NX2R0;$([NH]),$(N~[!#7])]=[NX2R0;$([NH]),$(N~[!#7])]

Filter18_oxime_ester:
[!#7 <https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>]C=NOC,S,P <http://%3Do/>* ->
[!#7 <https://github.com/rdkit/rdkit/issues/7>][CX3]=NOC,S,P <http://%3Do/>*
Adding explicit connections takes care of #1
<https://github.com/rdkit/rdkit/issues/1>

This is similar to the cases above, the revised SMARTS won't catch


[CX3;$(C[!#7]),$([CH2])]=NO[C,S,P]*

**Comment by PatWalters** [comment]
Sorry I hit send too soon.

acyclic imine:
***@***.***=[C!R,#1 <https://github.com/rdkit/rdkit/issues/1>]-C ->
***@***.***=[C!R]-C
hypervalent hydrogen dropped

I think the intent of the acyclic imine rule was to catch both and
secondary imines.   The revised rule doesn't catch primary imines.   This
seems to do the trick.  There may be a simpler way to do this.

***@***.***)

Filter5_azo:
[!#7 <https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>][NX2R0]=[NX2R0][!#7
<https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>] =>
[!#7 <https://github.com/rdkit/rdkit/issues/7>][NX2R0]=[NX2R0][!#7
<https://github.com/rdkit/rdkit/issues/7>]
X is explicit connections, don't need explicit hydrogen

In this case the intent appears to be to also catch things like
CN=N
Which would be missed by the revised pattern.  I think I'd write that one
as

[NX2R0;$([NH]),$(N~[!#7])]=[NX2R0;$([NH]),$(N~[!#7])]

Filter18_oxime_ester:
[!#7 <https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>]C=NOC,S,P <http://%3Do/>* ->
[!#7 <https://github.com/rdkit/rdkit/issues/7>][CX3]=NOC,S,P <http://%3Do/>*
Adding explicit connections takes care of #1
<https://github.com/rdkit/rdkit/issues/1>

This is similar to the cases above, the revised SMARTS won't catch
acyclic imine:
***@***.***=[C!R,#1 <https://github.com/rdkit/rdkit/issues/1>]-C ->
***@***.***=[C!R]-C
hypervalent hydrogen dropped

I think the intent of the acyclic imine rule was to catch both and
secondary imines.   The revised rule doesn't catch primary imines.   This
seems to do the trick.  There may be a simpler way to do this.

***@***.***)

Filter5_azo:
[!#7 <https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>][NX2R0]=[NX2R0][!#7
<https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>] =>
[!#7 <https://github.com/rdkit/rdkit/issues/7>][NX2R0]=[NX2R0][!#7
<https://github.com/rdkit/rdkit/issues/7>]
X is explicit connections, don't need explicit hydrogen

In this case the intent appears to be to also catch things like
CN=N
Which would be missed by the revised pattern.  I think I'd write that one
as

[NX2R0;$([NH]),$(N~[!#7])]=[NX2R0;$([NH]),$(N~[!#7])]

Filter18_oxime_ester:
[!#7 <https://github.com/rdkit/rdkit/issues/7>,#1
<https://github.com/rdkit/rdkit/issues/1>]C=NOC,S,P <http://%3Do/>* ->
[!#7 <https://github.com/rdkit/rdkit/issues/7>][CX3]=NOC,S,P <http://%3Do/>*
Adding explicit connections takes care of #1
<https://github.com/rdkit/rdkit/issues/1>

This is similar to the cases above, the revised SMARTS won't catch

C=NOCC

I think the SMARTS could be rewritten as

[CX3;$(C[!#7]),$([CH2])]=NO[C,S,P]*

Filter30_beta_halo_carbonyl:
[#6 <https://github.com/rdkit/rdkit/issues/6>,#1
<https://github.com/rdkit/rdkit/issues/1>]C(=O)CC[Br,I] ->
[$(C(-C)(-C)(=O)),$(CH <http://%3Do/>-C)](=O)CC[Br,I]
Make explicit recursive smarts for the #6
<https://github.com/rdkit/rdkit/issues/6>,#1
<https://github.com/rdkit/rdkit/issues/1>

This one looks correct to me although the SMARTS seem to have been mangled
somewhere in the GitHub to email chain.

**Comment by bp-kelley** [comment]
@PatWalters if you put the smarts patterns in markdown back ticks  they will format correctly.  Otherwise, send me the email directly and I'll format them.

**Comment by bp-kelley** [comment]
For the imine the smarts pattern:

```
 C-!@[NX2]=[C!R,#1]-C
```

```
for atom in smarts.GetAtoms(): 
  print(atom.DescribeQuery())
```

```
AtomType 6 = val

AtomAnd
  AtomType 7 = val
  AtomTotalDegree 2 = val

AtomOr
  AtomAnd
    AtomType 6 = val
    AtomInNRings -1 != val
  AtomAtomicNum 1 = val

AtomType 6 = val
```
Is saying you can double bound to a non ring carbon or a hydrogen.  The double bond to a hydrogen makes no sense here.

Daylight has the following for primary and secondary:

```
Substituted or un-substituted imine:
[$([CX3]([#6])[#6]),$([CX3H][#6])]=[$([NX2][#6]),$([NX2H])]
```

So I can switch it over that this I think.

**Comment by bp-kelley** [comment]
(I mean use this for the imine and add the acyclic)

**Comment by bp-kelley** [comment]
@PatWalters whoa, I was a bit brain dead there.  here is the current attempt, let me know if it works for you.  Some of those queries just weren't doing what was on the tin.


```
# filters_azo
smarts = "[$([NX2R0]-[!#7]),$([NX2R0H1])]=[$([NX2R0]-[!#7]),$([NX2R0H1])]"
tests = [("CN=NC", True), ("CN=N", True), ("N=N", True), ("NN=N", False)]
pat = Chem.MolFromSmarts(smarts)
pat = Chem.MergeQueryHs(pat)
for smi, res in tests:
    m = Chem.MolFromSmiles(smi)
    assert m.HasSubstructMatch(pat) == res
```


```
# filters Filter30_beta_halo_carbonyl
smarts = "[$(C(-C)(-C)(=O)),$([CH](=O)-C)](=O)CC[Br,I]"
tests = (("CC(=O)CCBr", True), ("CC(=O)CCI", True), ("C(=O)CCI", True), 
                  ("NC(=O)CCI", False))
pat = Chem.MolFromSmarts(smarts)
pat = Chem.MergeQueryHs(pat)
for s, result in tests:
    m = Chem.MolFromSmiles(s)
    assert m.HasSubstructMatch(pat) == result, s
```

```
# filters acyclic imine
# C-!@[NX2]=[C!R,#1]-C
smarts = "[$([CX3R0]([#6])[#6]),$([CX3HR0][#6])]=[$([NX2R0][#6]),$([NX2HR0])]"
tests = (("CN=C", False), ("CN=CC", True) , ("CN=C(C)", True), ("C1N=C1", False))
pat = Chem.MolFromSmarts(smarts)
pat = Chem.MergeQueryHs(pat)
for smi, res in tests:
    m = Chem.MolFromSmiles(smi)
    assert m.HasSubstructMatch(pat) == res, smi
```

```
# Filter18_oxime_ester: [!#7,#1]C=NO[C,S,P](=O)*
smarts = "[$([CX3H1]-[!#7]),$([CX3H2])]=NO[C,S,P](=O)*"
tests = (("CC=NOS(=O)N", True), ("NC=NOP(=O)N", False), ("C=NOP(=O)N", True))
pat = Chem.MolFromSmarts(smarts)
pat = Chem.MergeQueryHs(pat)
for smi, res in tests:
    m = Chem.MolFromSmiles(smi)
    assert m.HasSubstructMatch(pat) == res, (smi, res)
```

**Comment by bp-kelley** [comment]
@greglandrum this is ready for final review.  The final changes are documented here and have been added to the unit tests.

```
def test_smarts(sma,tests):
    pat = Chem.MolFromSmarts(sma)
    pat = Chem.MergeQueryHs(pat)
    for smi, res in tests:
        m = Chem.MolFromSmiles(smi)
        assert m.HasSubstructMatch(pat) == res, (smi, res)

# filters acyclic imine - LINT: C-!@[NX2]=[C!R,#1]-C
smarts = "[$([CX3R0]([#6])[#6]),$([CX3HR0][#6])]=[$([NX2R0][#6]),$([NX2HR0])]"
tests = (("CN=C", False), 
         ("CN=CC", True) , 
         ("CN=C(C)", True), 
         ("C1N=C1", False))
test_smarts(smarts, tests)

# filter5_azo
smarts = "[$([NX2R0]-[!#7]),$([NX2R0H1])]=[$([NX2R0]-[!#7]),$([NX2R0H1])]"
tests = [("CN=NC", True), 
         ("CN=N", True), 
         ("N=N", True), 
         ("NN=N", False)]
test_smarts(smarts, tests)

# Filter9_metal [!#1!#6!#7!#8!#9!#15!#16!#17!#35!#53]~[*,#1]
smarts = "[$([!#1!#6!#7!#8!#9!#15!#16!#17!#35!#53]~[*]),$([!#1!#6!#7!#8!#9!#15!#16!#17!#35!#53;h])]"
tests = [("[Fe]C", True),
         ("[FeH]", True),
         ("[Fe]", False)
        ]
test_smarts(smarts, tests)

# Filter12_nitroso [*,#1]~[NX2]~[OX1]
smarts = "[$([NX2]~*),$([NX2H])]~[OX1]"
tests = [("CN=O", True), 
         ("N=O", True)]
test_smarts(smarts, tests)

# Filter13_PS_double_bond S=P~[*,#1]
smarts = "S=[$([PX2]~*),$([PX2H])]"
tests = [("S=P", True), 
         ("S=PC", True)]
test_smarts(smarts, tests)

# Filter29_thioester CC(=[OX1,SX1,NH])[SX2][*,#1]
smarts = "CC(=[OX1,SX1,NH])[$([SX2]-*),$([SX2H])]"
tests = (("CC(=O)SC", True),
         ("CC(=S)SC", True),
         ("CC(=N)SC", True),
         ("CC(=O)S", True),
         ("CC(=S)S", True),
         ("CC(=N)S", True),
        )
test_smarts(smarts, tests)

# filters Filter30_beta_halo_carbonyl
smarts = "[$(C(-C)(-C)(=O)),$([CH](=O)-C)](=O)CC[Br,I]"
tests = (("CC(=O)CCBr", True), 
         ("CC(=O)CCI", True), 
         ("C(=O)CCI", True), 
         ("NC(=O)CCI", False))
test_smarts(smarts, tests)

# Filter18_oxime_ester: [!#7,#1]C=NO[C,S,P](=O)*
smarts = "[$([CX3H1]-[!#7]),$([CX3H2])]=NO[C,S,P](=O)*"
tests = (("CC=NOS(=O)N", True), 
         ("NC=NOP(=O)N", False), 
         ("C=NOP(=O)N", True))
test_smarts(smarts, tests)

#Filter31_so_bond [*,#1][SX2R0]O[*,#1]
smarts = "[SX2R0][OX2]"
tests = (("SO", True), 
         ("SOC", True), 
         ("c1csocc1", False))
test_smarts(smarts, tests)

#Filter32_oo_bond [*,#1][OR0]O[*,#1]
smarts = "[OX2R0][OX2]"
tests = (("OO", True), 
         ("COO", True), 
         ("c1coocc1", False))
test_smarts(smarts, tests)

#Filter44_michael_acceptor2 [#6,#1,OH0][C!$(C1(=O)C=CC(=O)C=C1)](=O)C=[C!$(C[Nv3X3,OH])]
smarts = "[$([C!$(C1(=O)C=CC(=O)C=C1)]C),$([Ch]),$(C[OH0])](=O)C=[C!$(C[Nv3X3,OH])]"
tests = (("CC(=O)C=C", True),
         ("C(=O)C=C", True),
         ("OC(=O)C=C", True),
         ("C1(=O)C=CC(=O)C=C1", False),
        )
test_smarts(smarts, tests)

#Filter49_halogen [*,#1]~[F,Cl,Br,I]~[*,#1]
# note, rdkit already doesn't accept FX2, ClX2 and BrX2
#  these are left for posterity
smarts = "[FX2,ClX2,BrX2,IX2,IX3,IX4,IX5]"
tests = (
         ("CIC", True),
         ("CI(C)C", True)
        )
test_smarts(smarts, tests)

# filter 89 hydroxylamine
smarts = "[*!$(C=O)]!@N!@[$([OX2])]"#"[*,#1]" 
tests = (("CC=NOS(=O)N", True), 
         ("NC=NO", True), 
         ("C(=O)NOS", False))
test_smarts(smarts, tests)
```

**Comment by greglandrum** [comment]
@bp-kelley kind of a basic question as I'm starting the review: why do we still have `Filters.cpp.in` at all? It doesn't seem to be used for anything and duplicates (not exactly) a bunch of code.

File: Code/GraphMol/FilterCatalog/Filters.cpp [hunk 1]
```diff
@@ -791,6 +791,906 @@ const FilterProperty_t ZINC_PROPS[] = {
 const unsigned int NUM_ZINC_PROPS =
     static_cast<unsigned int>(sizeof(ZINC_PROPS) / sizeof(FilterProperty_t));
 
+/////////////////////////////////////////////////////////////////////////////////////////
+// # Chembl 23 structural alert, rule set CHEMBL_Glaxo
+//
+const FilterData_t CHEMBL_Glaxo[] = {
```
Comment (greglandrum) [inline Code/GraphMol/FilterCatalog/Filters.cpp:797]: I think having these giant collections of SMARTS in this file makes it unmanageable.
I'd suggest going with:
`chembl_glaxo.in`, `chembl_dundee.in`, etc.
(yeah, I know we have other definitions in here too, but we could think about moving them into their own `.in` files in a separate PR

**Review by greglandrum — CHANGES_REQUESTED** [review]
I obviously didn't review the SMARTS themselves. ;-)

I think some reorganization of this, specifically pulling the new filter collections out into their own `.in` files, would help with maintainability.

Otherwise this looks fine.

File: Code/GraphMol/FilterCatalog/Filters.cpp (hunk 1 above)
Comment (bp-kelley) [reply Code/GraphMol/FilterCatalog/Filters.cpp:797]: I also noted that the Filters.cpp.in was unused.  I can remove it in this PR or we can merge this and make a separate one for splitting out the catalogs and removing it.

Comment (greglandrum) [reply Code/GraphMol/FilterCatalog/Filters.cpp:797]: > I also noted that the Filters.cpp.in was unused.  I can remove it in this PR or we can merge this and make a separate one for splitting out the catalogs and removing it.

I think it's worth removing it here and then doing a separate PR to pull the filters out

Comment (bp-kelley) [reply Code/GraphMol/FilterCatalog/Filters.cpp]: @greglandrum it was easy enough to split up.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by bp-kelley** [comment]
Thanks for the smarts review @PatWalters!

**Comment by greglandrum** [comment]
> Thanks for the smarts review @PatWalters!

yeah, thanks Pat.

**Comment by PatWalters** [comment]
Always an honor to contribute to the RDKit

**Pull request merged**